src/combinatorics/simple_graph/connectivity.lean

```diff
@@ -394,6 +399,228 @@ begin
   apply h.of_append_left,
 end
 
+/-! ### Walk decompositions -/
+
+section walk_decomp
+variables [decidable_eq V]
+
+/-- Given a vertex in the support of a path, give the path up until (and including) that vertex. -/
+def take_until : Π {v w : V} (p : G.walk v w) (u : V) (h : u ∈ p.support), G.walk v u
+| v w nil u h := by rw mem_support_nil_iff.mp h
+| v w (cons r p) u h :=
+  if hx : v = u
+  then by subst u
+  else cons r (take_until p _ $ h.cases_on (λ h', (hx h'.symm).elim) id)
+
+/-- Given a vertex in the support of a path, give the path from that vertex to the end. -/
+def drop_until : Π {v w : V} (p : G.walk v w) (u : V) (h : u ∈ p.support), G.walk u w
+| v w nil u h := by rw mem_support_nil_iff.mp h
+| v w (cons r p) u h :=
+  if hx : v = u
+  then by { subst u, exact cons r p }
+  else drop_until p _ $ h.cases_on (λ h', (hx h'.symm).elim) id
+
+/-- The `take_until` and `drop_until` functions split a walk into two pieces.
+The lemma `count_support_take_until_eq_one` specifies where this split occurs. -/
+@[simp]
+lemma take_spec {u v w : V} (p : G.walk v w) (h : u ∈ p.support) :
+  (p.take_until u h).append (p.drop_until u h) = p :=
+begin
+  induction p,
+  { rw mem_support_nil_iff at h,
+    subst u,
+    refl, },
+  { obtain (rfl|h) := h,
+    { simp! },
+    { simp! only,
+      split_ifs with h'; subst_vars; simp [*], } },
+end
+
+@[simp]
+lemma count_support_take_until_eq_one {u v w : V} (p : G.walk v w) (h : u ∈ p.support) :
+  (p.take_until u h).support.count u = 1 :=
+begin
+  induction p,
+  { rw mem_support_nil_iff at h,
+    subst u,
+    simp!, },
+  { obtain (rfl|h) := h,
+    { simp! },
+    { simp! only,
+      split_ifs with h'; rw eq_comm at h'; subst_vars; simp! [*, list.count_cons], } },
+end
+
+lemma count_edges_take_until_le_one {u v w : V} (p : G.walk v w) (h : u ∈ p.support) (x : V) :
+  (p.take_until u h).edges.count ⟦(u, x)⟧ ≤ 1 :=
+begin
+  induction p with u' u' v' w' ha p' ih,
+  { rw mem_support_nil_iff at h,
+    subst u,
+    simp!, },
+  { obtain (rfl|h) := h,
+    { simp!, },
+    { simp! only,
+      split_ifs with h',
+      { subst h',
+        simp, },
+      { rw [edges_cons, list.count_cons],
+        split_ifs with h'',
+        { rw sym2.eq_iff at h'',
+          obtain (⟨rfl,rfl⟩|⟨rfl,rfl⟩) := h'',
+          { exact (h' rfl).elim },
+          { cases p'; simp! } },
+        { apply ih, } } } },
+end
+
+lemma support_take_until_subset {u v w : V} (p : G.walk v w) (h : u ∈ p.support) :
+  (p.take_until u h).support ⊆ p.support :=
```

Comment:
Do we have notions of sub-walk? Should we? (It's entirely possible the answer is no!)

Reply:
Not yet, and I don't have any immediately plans for it.

The closest we have at the moment is that we can give an explicit `append` expression: `(p.take_until u h).append (p.drop_until u h) = p`